# Module Import Test Notebook

This notebook validates that the refactored modules load correctly.

In [1]:
#========================================
# Setup paths and notebook defaults
#========================================
from pathlib import Path
import sys

import numpy as np
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as pp
import seaborn as sns

BASE_DIR = Path.cwd().resolve().parent
DATA_DIR = BASE_DIR / "data"
INPUT_DIR = DATA_DIR / "input"


if str(BASE_DIR) not in sys.path:
    sys.path.insert(0, str(BASE_DIR))

print(f"BASE_DIR: {BASE_DIR}")

BASE_DIR: C:\Users\F.Turner\Documents\MwM Proof of Concept V2


In [2]:
from src.ingest import *
from src.aggregate import *
from src.combine import *
from src.output import *

print("Imports succeeded: ingest.py, aggregate.py, combine.py, and output.py are available.")

Imports succeeded: ingest.py, aggregate.py, combine.py, and output.py are available.


In [2]:
df = pd.read_csv(DATA_DIR / "output" / "kenya" / "tables" / "combined_child_level.csv")

In [3]:
df

,date_time,admin1,admin2,admin3,school,sex,age,grade,age_group,haldo_sel,haldo_literacy,_uuid
0,2023-06-12 11:40:00,8,800,80005,school01,1,7,0,6-10,0.604167,0.175000,1923914235876098121
1,2023-06-12 12:00:00,8,800,80005,school02,1,7,0,6-10,0.604167,0.175000,14316521119066352241
2,2023-06-07 10:54:00,9,902,90201,school03,1,7,2,6-10,0.837500,0.431818,8029136157369546520
3,2023-06-07 11:16:00,9,902,90201,school04,1,7,1,6-10,0.437500,0.431818,10855389975063662343
4,2023-06-08 09:56:00,9,902,90203,school05,1,8,0,6-10,0.500000,0.150000,17252436797090343061
...,...,...,...,...,...,...,...,...,...,...,...,...
1241,2023-05-31 11:25:00,7,700,70005,school09,0,10,2,6-10,0.700000,0.125000,12995168672598444935
1242,2023-05-07 10:41:00,7,700,70005,school01,0,10,2,6-10,0.757143,0.125000,13681566951212052960
1243,2023-05-31 11:02:00,7,700,70005,school02,0,10,2,6-10,0.800000,0.150000,18226929263901925735
1244,2023-05-31 11:02:00,7,700,70005,school03,0,10,2,6-10,0.800000,0.150000,8410737275105088741


In [9]:
ds_obs = build_dataset(file_path=str(INPUT_DIR / "lesson_obs.xlsx"))
ds_obs = aggregates(ds_obs)
combined = add_to_combined(ds_obs, output_dir=BASE_DIR / "data" / "output")

Form ID obs_001 added to log. Processed data will be saved as a new file.
Metadata added to log:
{'form_id': 'obs_001', 'survey_name': 'lesson_obs', 'level': 'teacher', 'first_submission': Timestamp('2024-10-03 08:08:00'), 'last_submission': Timestamp('2024-11-05 09:04:00'), 'rows': 20, 'country': 'kenya', '_uuid_cols': None}
Data imported from file. Kobo metadata functions will be skipped (question dataframe and choices dataframe). Only basic metadata will be logged.
UUID column already exists. Skipping UUID generation.
Available dimensions: ['date_time', 'admin3', 'school', 'sex']
Missing dimensions: ['admin0', 'admin1', 'admin2', 'location', 'protection_status', 'age', 'disability', 'grade', 'round', 'age_group']
Resolved columns: {'date_time': '_submission_time', 'admin0': None, 'admin1': None, 'admin2': None, 'admin3': 'community', 'location': None, 'protection_status': None, 'school': 'center', 'sex': 'gender', 'age': None, 'disability': None, 'grade': None, 'round': None, 'age_g

In [3]:
ds_aser = build_dataset(file_path=str(INPUT_DIR / "aser_sample.csv"))
ds_aser = aggregates(ds_aser)
combined_aser = add_to_combined(ds_aser, output_dir=BASE_DIR / "data" / "output")

Form ID aser_001 already exists in the log. Processed data will be added to the existing dataset.
Metadata added to log:
{'form_id': 'aser_001', 'survey_name': 'aser_assessment', 'level': 'child', 'first_submission': '2026-01-01 00:00:00', 'last_submission': '2026-01-03 00:00:00', 'rows': 200, 'country': 'kenya', '_uuid_cols': np.float64(nan)}
Data imported from file. Kobo metadata functions will be skipped (question dataframe and choices dataframe). Only basic metadata will be logged.
Dropped columns: ['child_name'] to protect child privacy
Available dimensions: ['date_time', 'admin1', 'location', 'protection_status', 'school', 'sex', 'age', 'disability', 'round', 'age_group']
Missing dimensions: ['admin0', 'admin2', 'admin3', 'grade']
Resolved columns: {'date_time': '_submission_time', 'admin0': None, 'admin1': 'region', 'admin2': None, 'admin3': None, 'location': 'location', 'protection_status': 'protection_status', 'school': 'school_name', 'sex': 'child_sex', 'age': 'child_age', 'd

In [4]:
ds_haldo = build_dataset(file_path=str(INPUT_DIR / "haldo_sample.xlsx"))
ds_haldo = aggregates(ds_haldo)
combined_haldo = add_to_combined(ds_haldo, output_dir=BASE_DIR / "data" / "output")

Form ID haldo_001 added to log. Processed data will be saved as a new file.
Metadata added to log:
{'form_id': 'haldo_001', 'survey_name': 'haldo_assessment', 'level': 'child', 'first_submission': Timestamp('2023-04-11 11:40:00'), 'last_submission': Timestamp('2023-06-13 12:27:00'), 'rows': 1246, 'country': 'kenya', '_uuid_cols': None}
Data imported from file. Kobo metadata functions will be skipped (question dataframe and choices dataframe). Only basic metadata will be logged.
Available dimensions: ['date_time', 'admin1', 'admin2', 'admin3', 'school', 'sex', 'age', 'grade', 'age_group']
Missing dimensions: ['admin0', 'location', 'protection_status', 'disability', 'round']
Resolved columns: {'date_time': '_submission_time', 'admin0': None, 'admin1': 'governorate', 'admin2': 'district', 'admin3': 'subdistrict', 'location': None, 'protection_status': None, 'school': 'school', 'sex': 'sex', 'age': 'age', 'disability': None, 'grade': 'grade', 'round': None, 'age_group': 'age'}
Derived dime

In [5]:
ds_school = build_dataset(file_path=str(INPUT_DIR / "school_sample.csv"))
ds_school = aggregates(ds_school)
ds_school_combined = school_combined(ds_school, output_dir=BASE_DIR / "data" / "output")

Form ID school_001 added to log. Processed data will be saved as a new file.
Metadata added to log:
{'form_id': 'school_001', 'survey_name': 'school_summary', 'level': 'school', 'first_submission': Timestamp('2026-01-01 00:00:00'), 'last_submission': Timestamp('2026-03-01 00:00:00'), 'rows': 27, 'country': 'kenya', '_uuid_cols': None}
Data imported from file. Kobo metadata functions will be skipped (question dataframe and choices dataframe). Only basic metadata will be logged.
UID depends on 7 columns - may be unstable
Available dimensions: ['date_time', 'admin1', 'admin2', 'school']
Missing dimensions: ['admin0', 'admin3', 'location', 'protection_status', 'sex', 'age', 'disability', 'grade', 'round', 'age_group']
Resolved columns: {'date_time': '_submission_time', 'admin0': None, 'admin1': 'region', 'admin2': 'district', 'admin3': None, 'location': None, 'protection_status': None, 'school': 'school_name', 'sex': None, 'age': None, 'disability': None, 'grade': None, 'round': None, 'age

C:\Users\F.Turner\Documents\MwM Proof of Concept V2\src\combine.py:251: UserWarning: Converting to PeriodArray/Index representation will drop timezone information.
  .groupby(["school", teacher_combined["date_time"].dt.to_period("M")], dropna=False)


In [9]:
ds_school_combined.df_short.head(30)

,school,date_time,admin1,admin2,pta_active,pta_active,school_management_committee_active,plc_frequency,plc_frequency,safeguarding_policy,...,disruption_level,literacy_practice,numeracy_practice,sel_practice,student_teacher_ratio,learning_resources,gender_responsiveness,safe_environment,rtl_use,lesson_preparedness
0,school01,NaT,West,Zeta,no,no,yes,1.0,1.0,0.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,school01,2024-10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,3.000000,1.666667,2.0,1.666667,3.000000,2.333333,2.333333,2.666667,2.333333,2.333333
2,school01,2026-01,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,school02,NaT,West,Delta,no,no,yes,1.0,1.0,0.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,school02,2024-10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2.333333,2.333333,3.0,2.333333,2.666667,1.666667,2.666667,1.666667,2.333333,2.666667
5,school02,2026-01,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,school03,NaT,East,Gamma,yes,yes,no,1.0,1.0,1.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7,school03,2024-10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2.500000,1.500000,3.0,2.000000,1.000000,2.000000,2.500000,3.000000,3.000000,2.000000
8,school03,2026-01,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
9,school04,NaT,South,Alpha,yes,yes,yes,1.0,1.0,1.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Targeted Optimisation Plan

Cache config classification lookups (highest ROI, lowest risk)
Problem: each chart pair calls classification twice, and each call re-opens JSON files in report.py:55 and report.py:57, triggered from app.py:94 and app.py:95.
Change: load config once per level via cached helper (for example with st.cache_data) and classify using in-memory dicts.
Impact: removes repeated disk I/O across every rerun and every indicator-dimension pair.

Render only the currently selected level, not all three
Problem: all three tabs execute render paths in app.py:131, so one filter change recalculates Child + Teacher + School.
Change: replace level tabs with a level selector, or gate computation so only the active level runs.
Impact: often cuts dashboard compute by about 2/3 on each interaction.

Add a chart cap per indicator and lazy detail expansion
Problem: nested loops in app.py:83 and app.py:87 can generate many figures.
Change: show top N dimensions by default (for example 4-6), with an expander for “show all dimensions”.
Impact: large drop in Plotly construction/serialization time while preserving full detail on demand.

Precompute filtered dimension options once per rerun
Problem: option extraction and sorting in app.py:75 runs for each dimension every rerun.
Change: build an options map once from base df (or cached by country/level), then reuse in multiselects.
Impact: improves sidebar responsiveness, especially with high-cardinality columns.

Avoid repeated dataframe copies in hot loops
Problem: repeated slicing/copy/dropna in app.py:90 and conversion blocks around app.py:98 and app.py:116.
Change: do type normalization once per level before chart loops, and avoid unnecessary copy() unless mutation is required.
Impact: reduces memory churn and Python overhead.

Cache figure-ready aggregates for common plot paths
Problem: repeated groupby and crosstab in report.py:192, report.py:199, report.py:219.
Change: cache grouped outputs keyed by (country, level, indicator, dimension, filter_signature).
Impact: big gains when users tweak a subset of filters repeatedly.

Reduce per-figure axis/tick work
Problem: tick text rebuilding for every figure in report.py:251.
Change: only apply custom tick formatting for categorical axes with large labels; skip for numeric/clean short labels.
Impact: modest but noticeable when many charts are shown.

Keep data loading cached but tighten cache keys
Good: app.py:32 already caches load.
Improve: ensure cache depends only on (level, country) and clear cache only after ingestion (already done at app.py:184 and app.py:228).
Impact: avoids accidental unnecessary reloads.

Suggested rollout order (minimal structural change)

- Config caching in classification.
- Single active level rendering.
- Chart cap + optional “show all”.
- Loop/dataframe micro-optimizations.
- Optional aggregate caching if still slow.

How to validate each step

Add timing logs around app.py:54 and inside chart loop near app.py:87.
Track:
total rerun time
number of charts rendered
average time per chart
Compare baseline vs after each step, keeping same country/filter state.
If you want, I can draft a concrete patch plan with exact function signatures and cache-key structure before any code changes.